posts/2018-04-13-rethinking-monaderror.md

```diff
@@ -0,0 +1,456 @@
+
+Now that we figured out the shape, let's see what we can actually do with it.
+For throwing errors, we'll create a `raiseError` function that should return a value inside `F`, as it will obviously be able to yield an error.
+
+```scala
+trait MonadBlunder[F[_], G[_], E] {
+  def raiseError[A](e: E): F[A]
+}
+```
+
+This definition looks identical to the one defined one `MonadError` so let's move on to error-handling.
+For handled errors, we want to return a value inside `G`, so our `handleErrorWith` function should indeed return a `G[A]`:
+
+```scala
+trait MonadBlunder[F[_], G[_], E] {
+  ...
+
+  def handleErrorWith[A](fa: F[A])(f: E => F[A]): G[A]
+}
+```
+
+Looks good so far, right? 
+Well, we still have the problem that `f` might return an erronous value, so if we want to guarantee that the result won't have any errors, we'll have to change that to `G[A]` as well:
+
+```scala
+trait MonadBlunder[F[_], G[_], E] {
+  ...
+
+  def handleErrorWith[A](fa: F[A])(f: E => G[A]): G[A]
+}
+```
+
+And now we're off to a pretty good start, we fixed one short coming of `MonadError` with this approach.
+
+Another approach, maybe more obvious to some, might be to require the type constructor to take two arguments, one for the value and one for the error type.
+Let's see if we can define `raiseError` on top of it:
+
+```scala
+trait MonadBlunder[F[_, _], E] {
+  def raiseError[A](e: E): F[E, A]
+
+  ...
+}
+```
+
+This looks pretty familiar to what we already have, though now we have the guarantee that our type doesn't actually "hide" the error-type somewhere.
+Next up is `handleErrorWith`. Ideally after we handled the error we should again get back a type that signals that it doesn't have any errors. 
+We can do exactly that by choosing an unhabited type like `Nothing` as our error-type:
+
+
+```scala
+trait MonadBlunder[F[_, _], E] {
+  ...
+
+  def handleErrorWith[A](fa: F[E, A])(f: A => F[Nothing, A]): F[Nothing, A]
+}
+```
+
+And this approach works as well, however now we've forced the two type parameter shape onto implementors. This `MonadBlunder` has the following kind `(* -> * -> *) -> * -> *`.
+This means we can very easily define instances for types with two type parameters like `Either`.
+However, one issue might be that it's much easier to fit a type with two type parameters onto a type class that expects a single type constructor `(* -> *)` than to do it the other way around.
+
+For example try to implement the above `MonadBlunder[F[_, _], E]` for the standard `cats.effect.IO`.
+It's not going to be simple, whereas with the current encoding we can easily encode both `Either` and `IO`. 
+
+Next we're going to look at laws we can define to make sense of the behaviour we want.
+The first two laws should be fairly obvious. 
+If we `flatMap` over a value created by `raiseError` it shouldn't propogate:
+
+```scala
+def raiseErrorStops(e: E, a: A): Boolean =
+  F.raiseError[A](e).flatMap(_ => a.pure[F]) === F.raiseError[A](e)
```

Comment:
This is guaranteed by parametricity. So there's no need to test it.

Reply:
Hmmm, I think you're right, I adapted the law from here: https://github.com/purescript/purescript-transformers/blob/master/src/Control/Monad/Error/Class.purs#L18

Reply:
I.e. you can say that of the function given to `flatMap` can never be called due to there being no value of `A`, that's one side, but this law also requires that the `flatMap` call does absolutely nothing at all, which is a very slightly different statement.
I'm not sure I like it tbh, I'm going to have to take a better look. Seems Haskell doesn't have laws for `MonadError` at all.

Reply:
`flatMap` has to do nothing anyway, because of the law that states `flatMap(_.pure[F])` does nothing, as well as because `flatMap` can be specified in terms of `map` and `join`.

Reply:
Good point, I wonder if Phil thought of that or had something else in mind.